reader_concurrency_semaphore.cc

```diff
@@ -907,6 +907,8 @@ future<> reader_concurrency_semaphore::execution_loop() noexcept {
                 co_await coroutine::maybe_yield();
             }
         }
+
+        maybe_admit_waiters();
```

Comment:
Not sure what you mean. I added this to after the loop exhausts the ready list, just before it starts waiting on the conditional variable. This check ensures that we only go to sleep if no permit can be admitted.

Reply:
Suppose the condition that allows us to admit a read becomes true only after this new check. So we have this sequence:

 - a read gets queued
 - this check tries to admit it, but fails
 - the condition that allows admitting the read becomes true
 - we're still asleep

Reply:
`_ready_list` is always non-empty when the execution loop runs. If it is empty, the loop waits on the condition variable. The expectation is that by the time the current read finishes running, there will be another read pushed onto the `_ready_list`, via calls to `maybe_admit_waiters()`, called from the current read releasing resources. If there is no new read on the list, the loop waits on the condition variable. The current patch adds a forced poll for ready reads just before the loop goes to sleep, in case the previous read didn't trigger any (although that is very unlikely, but not impossible).

Reply:
How is it possible, barring a bug?

Reply:
I guess there's a bug in how the condition variable is used. Waits should always be in a look with a check just before, or using the overload of when() that accepts a predicate.

Reply:
But I think the code as written (before this patch) is equivalent. If there is a spurious wakeup, the while () loop will ignore it.

I still don't understand this patch.

Reply:
> How is it possible, barring a bug?

I don't know how this could happen in the real world. This patch is quite speculative to be honest. I could write a unit test reproducer but it is quite artificial. After all, what real read could complete a page and not release any resources? The only case I could think of is a very last page, where the page turns out to be empty and nothing was read. But in this case the reader (and the permit with it) would be destroyed, which releases the base resources of the permit.